### PR TITLE
TFM: configurable memory layout

### PR DESCRIPTION
**Pull request by oyvindronningstad**
~Depends on the upmerge~

### CODE REVIEW

File: subsys/tfm/boards/board/RTE_Device.h [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+// <e> USART (Universal synchronous - asynchronous receiver transmitter) [Driver_USART1]
+// <i> Configuration settings for Driver_USART1 in component ::Drivers:USART
+#define   RTE_USART1                    1
+//   <h> Pin Selection (0xFFFFFFFF means Disconnected)
+#if defined(CONFIG_BOARD_NRF5340PDK_NRF5340_CPUAPPNS) || defined(CONFIG_BOARD_NRF5340DK_NRF5340_CPUAPPNS)
+//     <o> TXD
+#define   RTE_USART1_TXD_PIN            25 // TODO: Add to devicetree
+//     <o> RXD
+#define   RTE_USART1_RXD_PIN            26 // TODO: Add to devicetree
+#else
```
Comment (oyvindronningstad) [inline subsys/tfm/boards/board/RTE_Device.h:53]: @anangl This UART is not present in the devicetree. Is there any chance of adding it?

File: samples/tfm/tfm_psa_test/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (hakonfam) [inline samples/tfm/tfm_psa_test/CMakeLists.txt:1]: Wrong license text?

File: samples/tfm/tfm_psa_test/app.overlay [hunk 3]
```diff
@@ -0,0 +1,9 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
```
Comment (hakonfam) [inline samples/tfm/tfm_psa_test/app.overlay:2]: 2021?

File: samples/tfm/tfm_psa_test/sample.yaml [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+        - "TOTAL FAILED    : 0"
+sample:
+  name: "TFM PSA Test"
+
+tests:
+  sample.tfm.psa_storage_test:
+    tags: tfm
+    extra_args: "CONFIG_TFM_PSA_TEST_STORAGE=y"
+    timeout: 100
+  # sample.tfm.psa_crypto_test:
```
Comment (hakonfam) [inline samples/tfm/tfm_psa_test/sample.yaml:19]: Work in progress or should this be removed?

File: samples/tfm/tfm_psa_test/src/main.c [hunk 5]
```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright (c) 2012-2014 Wind River Systems, Inc.
```
Comment (hakonfam) [inline samples/tfm/tfm_psa_test/src/main.c:2]: Is this correct? If yes, can we have it in the sdk-nrf repo? Unsure.

File: subsys/partition_manager/pm.yml.tfm [hunk 6]
```diff
@@ -19,8 +15,13 @@ tfm_primary:
 app_primary:
   span: [app]
 
+#ifndef CONFIG_TFM_BL2_FALSE
```
Comment (hakonfam) [inline subsys/partition_manager/pm.yml.tfm:18]: What if the default value is to disable MCUBoot? Not a comment directly to this PR, but I guess this is motivation enough to stop using the "use whatever is default TFM" as an option for different TFM kconfigs.

File: subsys/partition_manager/pm.yml.tfm [hunk 7]
```diff
@@ -29,8 +30,9 @@ tfm_secondary:
 app_secondary:
   placement: {after: [tfm_secondary]}
   share_size: app_primary
+#endif
 
 tfm_extra:
-  placement: {after: [app_secondary]}
+  placement: {after: [app_secondary, app]}
```
Comment (hakonfam) [inline subsys/partition_manager/pm.yml.tfm:36]: Is this documented in the commit message? Why is this here?

File: subsys/partition_manager/CMakeLists.txt [hunk 8]
```diff
@@ -78,7 +78,16 @@ if (CONFIG_BUILD_WITH_TFM)
     set_property(GLOBAL PROPERTY
       bl2_PM_HEX_FILE ${CMAKE_BINARY_DIR}/tfm/bin/bl2.hex
       )
+    set_property(GLOBAL PROPERTY
```
Comment (hakonfam) [inline subsys/partition_manager/CMakeLists.txt:81]: This (everything itside the `if (CONFIG_BUILD_WITH_TFM)` should really all be moved out to some TFM specific subsys folder CMakeLists.txt. Its not something you work on directly in this PR, but it would be great if you could handle it here.

File: subsys/tfm/boards/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,39 @@
+#-------------------------------------------------------------------------------
+# Copyright (c) 2020, Nordic Semiconductor ASA.
```
Comment (hakonfam) [inline subsys/tfm/boards/CMakeLists.txt:2]: 2021?

Comment (hakonfam) [reply subsys/tfm/boards/CMakeLists.txt:2]: I'm actually not sure, since this is written in 2020 (or is it? :laughing: )

File: subsys/tfm/boards/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+#-------------------------------------------------------------------------------
+
+# NOTE: All code in this folder and subfolders is run as part of the TF-M
+#       build, not in a Zephyr context.
+
+cmake_policy(SET CMP0076 NEW)
```
Comment (hakonfam) [inline subsys/tfm/boards/CMakeLists.txt:11]: Should we comment why this is here?

File: subsys/tfm/boards/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+set(partition_includes
+  partition
+  ${CMAKE_BINARY_DIR}/../zephyr/include/generated
+  )
+
+set(board_includes
+  board
+  ${CMAKE_BINARY_DIR}/../zephyr/misc/generated/syscalls_links/include
+  )
+
```
Comment (hakonfam) [inline subsys/tfm/boards/CMakeLists.txt:24]: nit: 2 newlines

**Review by hakonfam — COMMENTED** [review]
Should we have 2020 or 2021 on these file? 
Should they be in subsys when they are only used by the tfm project? What about `ext` or somewhere else?

It feels weird putting stuff which is only used by a different project inside subsys, but maybe its correct.

File: subsys/partition_manager/pm.yml.tfm (hunk 6 above)
Comment (oyvindronningstad) [reply subsys/partition_manager/pm.yml.tfm:18]: Yes, we have a task for changing that. Until then, IMO we can assume that this doesn't change, i.e. default TFM -> BL2 is enabled.

File: subsys/partition_manager/pm.yml.tfm (hunk 7 above)
Comment (oyvindronningstad) [reply subsys/partition_manager/pm.yml.tfm:36]: Because sometimes app_secondary is not part of the build (if BL2 is disabled)

File: subsys/tfm/boards/CMakeLists.txt (hunk 9 above)
Comment (oyvindronningstad) [reply subsys/tfm/boards/CMakeLists.txt:2]: will change

File: subsys/tfm/boards/CMakeLists.txt (hunk 10 above)
Comment (oyvindronningstad) [reply subsys/tfm/boards/CMakeLists.txt:11]: Looks like it is added to avoid a Cmake warning. All TFM platforms set this. I can add a comment if you want it.

**Review by andvib — APPROVED** [review]

File: subsys/tfm/boards/CMakeLists.txt (hunk 9 above)
Comment (oyvindronningstad) [reply subsys/tfm/boards/CMakeLists.txt]: Changed

File: subsys/tfm/boards/CMakeLists.txt (hunk 10 above)
Comment (oyvindronningstad) [reply subsys/tfm/boards/CMakeLists.txt]: Removed

File: subsys/tfm/boards/CMakeLists.txt (hunk 11 above)
Comment (oyvindronningstad) [reply subsys/tfm/boards/CMakeLists.txt:23]: Fixed

File: subsys/partition_manager/pm.yml.tfm (hunk 7 above)
Comment (oyvindronningstad) [reply subsys/partition_manager/pm.yml.tfm:36]: Does this need a comment @hakonfam?

File: subsys/partition_manager/CMakeLists.txt (hunk 8 above)
Comment (oyvindronningstad) [reply subsys/partition_manager/CMakeLists.txt:81]: Moved to subsys/tfm/CMakeLists.txt

File: subsys/tfm/boards/nrf5340_cpuapp_eng_a/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+#-------------------------------------------------------------------------------
+
+cmake_policy(SET CMP0076 NEW)
+set(CMAKE_CURRENT_SOURCE_DIR ${CMAKE_CURRENT_LIST_DIR})
+set(NRF_BOARD_SELECTED True)
+
+add_subdirectory(../../../../../modules/tee/tfm/trusted-firmware-m/platform/ext/target/nordic_nrf/common/nrf5340 nrf5340)
```
Comment (tejlmand) [inline subsys/tfm/boards/nrf5340_cpuapp_eng_a/CMakeLists.txt:12]: We should do something with this. my eyes are hurting me.

**Review by tejlmand — CHANGES_REQUESTED** [review]

File: subsys/tfm/boards/nrf5340_cpuapp_eng_a/CMakeLists.txt (hunk 12 above)
Comment (tejlmand) [reply subsys/tfm/boards/nrf5340_cpuapp_eng_a/CMakeLists.txt:12]: how about:
```suggestion
add_subdirectory(${Trusted\ Firmware\ M_SOURCE_DIR}/platform/ext/target/nordic_nrf/common/nrf5340 nrf5340)
```

File: subsys/tfm/boards/nrf5340_cpuapp/CMakeLists.txt [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+#-------------------------------------------------------------------------------
+
+cmake_policy(SET CMP0076 NEW)
+set(CMAKE_CURRENT_SOURCE_DIR ${CMAKE_CURRENT_LIST_DIR})
+set(NRF_BOARD_SELECTED True)
+
+add_subdirectory(../../../../../modules/tee/tfm/trusted-firmware-m/platform/ext/target/nordic_nrf/common/nrf5340 nrf5340)
```
Comment (tejlmand) [inline subsys/tfm/boards/nrf5340_cpuapp/CMakeLists.txt:12]:
```suggestion
add_subdirectory(${Trusted\ Firmware\ M_SOURCE_DIR}/platform/ext/target/nordic_nrf/common/nrf5340 nrf5340)
```

File: subsys/tfm/boards/nrf9160/CMakeLists.txt [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+#-------------------------------------------------------------------------------
+
+cmake_policy(SET CMP0076 NEW)
+set(CMAKE_CURRENT_SOURCE_DIR ${CMAKE_CURRENT_LIST_DIR})
+set(NRF_BOARD_SELECTED True)
+
+add_subdirectory(../../../../../modules/tee/tfm/trusted-firmware-m/platform/ext/target/nordic_nrf/common/nrf5340 nrf5340)
```
[the remark this answers is not in the document]
Comment (tejlmand) [reply subsys/tfm/boards/nrf9160/CMakeLists.txt:12]: or even better:
```suggestion
add_subdirectory(${Trusted\ Firmware\ M_SOURCE_DIR}/platform/ext/target/nordic_nrf/common/nrf9160 nrf9160)
```

File: subsys/tfm/boards/nrf9160/preload.cmake [hunk 15]
```diff
@@ -0,0 +1,10 @@
+#-------------------------------------------------------------------------------
+# Copyright (c) 2021, Nordic Semiconductor ASA.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+#-------------------------------------------------------------------------------
+
+set(NRF_PATH platform/ext/target/nordic_nrf)
```
Comment (tejlmand) [inline subsys/tfm/boards/nrf9160/preload.cmake:8]: To make it completely clear to the user where this file is located (that is outside this repo), I would suggest:
```suggestion
set(TFM_NRF_DIR ${Trusted\ Firmware\ M_SOURCE_DIR}/platform/ext/target/nordic_nrf)
```
Using `DIR` as that is the common name for a path in CMake.
Also suggesting `TFM_NRF_DIR` to not confuse readers with the `NRF_DIR` in this repo.
https://github.com/nrfconnect/sdk-nrf/blob/master/CMakeLists.txt#L8

File: subsys/tfm/boards/nrf9160/preload.cmake [hunk 16]
```diff
@@ -0,0 +1,10 @@
...
+
+include(${NRF_PATH}/common/nrf9160/preload.cmake)
```
Comment (tejlmand) [inline subsys/tfm/boards/nrf9160/preload.cmake:10]: as continuation of the other comment:
```suggestion
include(${TFM_NRF_DIR}/common/nrf9160/preload.cmake)
```

**Review by tejlmand — CHANGES_REQUESTED** [review]
couple of more comments

File: subsys/partition_manager/Kconfig [hunk 17]
```diff
@@ -174,6 +174,15 @@ config PM_PARTITION_SIZE_TFM_EXTRA
 
 endif
 
+config TFM_BOARD
+	string
+	default "../../../../../../../nrf/subsys/tfm/boards/nrf9160" if SOC_NRF9160_SICA
```
Comment (tejlmand) [inline subsys/partition_manager/Kconfig:179]: we should not use relative paths in Kconfig, cause it is never clear what it is relative to.
Sometimes people think relative to a conf file, sometimes to app dir, and in this case I assume you consider relative to a folder inside the tfm repo going into nrf, making thing break if we relocate the TFM module later (as example going from 3 -> single repo clone or switching between nrf/zephyr fork and vanilla upstream).

But relative is very unclear when you start placing settings in Kconfig fragments, or specify them using `-DCONFIG_<foo>`.

Why does this value need to be a path in Kconfig ?
Could we not keep the upstream symbol, and then in CMake, adjust the value of `CONFIG_TFM_BOARD` like:
```
set(CONFIG_TFM_BOARD ${OUR_PREFIX_WITH_PATH}/${CONFIG_TFM_BOARD})
```
would of course require that we put out boards in `nordic_nrf/nrf9160dk_nrf9160` instead of just `nrf9160`.

**Review by tejlmand — CHANGES_REQUESTED** [review]

File: subsys/partition_manager/Kconfig (hunk 17 above)
Comment (tejlmand) [reply subsys/partition_manager/Kconfig:179]: As example, you may do:
```
file(RELATIVE_PATH
     TFM_NRF_RELATIVE_TARGET_DIR
     ${ZEPHYR_TFM_MODULE_DIR}/trusted-firmware-m/platform/ext/target/
     ${NRF_DIR}/subsys/tfm/boards
)
set(CONFIG_TFM_BOARD ${TFM_NRF_RELATIVE_TARGET_DIR}/${CONFIG_TFM_BOARD})
```

which will set: `TFM_NRF_RELATIVE_TARGET_DIR` to `../../../../../../../nrf/subsys/tfm/boards`

File: subsys/tfm/boards/nrf9160/CMakeLists.txt (hunk 14 above)
[the remark this answers is not in the document]
Comment (oyvindronningstad) [reply subsys/tfm/boards/nrf9160/CMakeLists.txt:12]: is `${Trusted\ Firmware\ M_SOURCE_DIR}` defined somewhere or should I define it myself?

File: CODEOWNERS [hunk 18]
```diff
@@ -119,6 +119,7 @@ Kconfig*                                  @tejlmand
 /subsys/shell/                            @nordic-krch
 /subsys/spm/                              @ioannisg
 /subsys/nonsecure/                        @ioannisg
+/subsys/tfm/                              @oyvindronningstad @ioannisg @hakonfam
```
[the remark this answers is not in the document]
Comment (oyvindronningstad) [reply CODEOWNERS:122]: wouldn't that be confusing? Since it won't really be its own module, it's just supplementing another module that is defined somewhere else.

File: subsys/tfm/boards/nrf5340_cpuapp/CMakeLists.txt [hunk 19]
```diff
@@ -0,0 +1,14 @@
+#-------------------------------------------------------------------------------
+# Copyright (c) 2021, Nordic Semiconductor ASA.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+#-------------------------------------------------------------------------------
+
+cmake_policy(SET CMP0076 NEW)
+set(CMAKE_CURRENT_SOURCE_DIR ${CMAKE_CURRENT_LIST_DIR})
```
[the remark this answers is not in the document]
Comment (oyvindronningstad) [reply subsys/tfm/boards/nrf5340_cpuapp/CMakeLists.txt]: Removed

File: subsys/tfm/boards/nrf5340_cpuapp_eng_a/CMakeLists.txt [hunk 20]
```diff
@@ -0,0 +1,14 @@
+#-------------------------------------------------------------------------------
+# Copyright (c) 2021, Nordic Semiconductor ASA.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+#-------------------------------------------------------------------------------
+
+cmake_policy(SET CMP0076 NEW)
+set(CMAKE_CURRENT_SOURCE_DIR ${CMAKE_CURRENT_LIST_DIR})
```
[the remark this answers is not in the document]
Comment (oyvindronningstad) [reply subsys/tfm/boards/nrf5340_cpuapp_eng_a/CMakeLists.txt]: Removed

File: subsys/tfm/boards/nrf9160/CMakeLists.txt [hunk 21]
```diff
@@ -0,0 +1,14 @@
+#-------------------------------------------------------------------------------
+# Copyright (c) 2021, Nordic Semiconductor ASA.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+#-------------------------------------------------------------------------------
+
+cmake_policy(SET CMP0076 NEW)
+set(CMAKE_CURRENT_SOURCE_DIR ${CMAKE_CURRENT_LIST_DIR})
```
[the remark this answers is not in the document]
Comment (oyvindronningstad) [reply subsys/tfm/boards/nrf9160/CMakeLists.txt]: Removed

File: subsys/tfm/boards/nrf9160/CMakeLists.txt (hunk 14 above)
[the remark this answers is not in the document]
Comment (oyvindronningstad) [reply subsys/tfm/boards/nrf9160/CMakeLists.txt]: > perhaps ?

Yep, thanks :). Fixed

[the remark this answers is not in the document]
Comment (oyvindronningstad) [reply subsys/tfm/boards/nrf9160/CMakeLists.txt]: > is `${Trusted\ Firmware\ M_SOURCE_DIR}` defined somewhere or should I define it myself?

@tejlmand ?

File: subsys/tfm/boards/nrf9160/preload.cmake (hunk 15 above)
Comment (oyvindronningstad) [reply subsys/tfm/boards/nrf9160/preload.cmake]: It was only used here, so I removed the variable.

File: subsys/tfm/boards/nrf9160/preload.cmake (hunk 16 above)
Comment (oyvindronningstad) [reply subsys/tfm/boards/nrf9160/preload.cmake]: Removed

File: subsys/partition_manager/Kconfig (hunk 17 above)
Comment (oyvindronningstad) [reply subsys/partition_manager/Kconfig]: @tejlmand I created a new Kconfig TFM_BOARD_BASE_DIR. Working with variables in Cmake gets tricky because you don't know quite which order things will happen in. I opted for KConfig string that can contain Cmake variables. This means that the Cmake context is known (TFM module), but the value can still be overridden from anywhere. Tell me what you think

File: subsys/tfm/boards/nrf9160/CMakeLists.txt (hunk 14 above)
[the remark this answers is not in the document]
Comment (tejlmand) [reply subsys/tfm/boards/nrf9160/CMakeLists.txt]: it's defined by Trusted Firmware M itself.
It's a CMake controlled variable, that gets set when running `project()` which means this variable is defined with this line:
https://github.com/oyvindronningstad/sdk-trusted-firmware-m/blob/plat-ext-2/trusted-firmware-m/CMakeLists.txt#L64
```
project("Trusted Firmware M" VERSION ${TFM_VERSION} LANGUAGES C ASM)
```

File: samples/tfm/tfm_psa_test/CMakeLists.txt (hunk 2 above)
Comment (oyvindronningstad) [reply samples/tfm/tfm_psa_test/CMakeLists.txt]: Fixed in #3728

File: samples/tfm/tfm_psa_test/app.overlay (hunk 3 above)
Comment (oyvindronningstad) [reply samples/tfm/tfm_psa_test/app.overlay]: Fixed in #3728

File: samples/tfm/tfm_psa_test/sample.yaml (hunk 4 above)
Comment (oyvindronningstad) [reply samples/tfm/tfm_psa_test/sample.yaml]: Work in progress. I'd like to keep it until these suites work (depends on @frkv's work)

File: samples/tfm/tfm_psa_test/src/main.c (hunk 5 above)
Comment (oyvindronningstad) [reply samples/tfm/tfm_psa_test/src/main.c]: Fixed in #3728

File: CODEOWNERS (hunk 18 above)
[the remark this answers is not in the document]
Comment (tejlmand) [reply CODEOWNERS:122]: > wouldn't that be confusing?

in my opinion it is more confusing to have tfm code under `subsys`.
Code not used in the Zephyr build system, but instead is used inside the tfm build system as an external project.

> Since it won't really be its own module, it's just supplementing another module that is defined somewhere else.

This is exactly glue code for the TFM module. Not glue code to integrate TFM into Zephyr build system, but glue code to integrate NCS repo into TFM build system. So just in the opposite direction.
And if we want in future, we could have a NCS specific `modules/tfm/CMakeLists.txt` that would be used as glue code, where we could even adjust what variables are fed to TFM build, and that way handle the path much cleaner.
But that would have to wait for: https://github.com/zephyrproject-rtos/zephyr/pull/30905

[the remark this answers is not in the document]
Comment (oyvindronningstad) [reply CODEOWNERS:122]: Fair point, I'll move it. But how about the code now in subsys/tfm/CMakeLists.txt and subsys/tfm/Kconfig?

[the remark this answers is not in the document]
Comment (oyvindronningstad) [reply CODEOWNERS:122]: Well maybe not. I'll move that as well.

File: subsys/partition_manager/Kconfig (hunk 17 above)
Comment (tejlmand) [reply subsys/partition_manager/Kconfig]: >  @tejlmand I created a new Kconfig TFM_BOARD_BASE_DIR.

Thanks for the proposal. I see it requires an additional patch upstream, and in that case, I think it would be even better if TFM itself could be improved to allow out of tree boards.
This setting is similar to a board root in Zephyr, which is something we normally don't put in Kconfig, but of course it is a little different here, as this is a Zephyr module.

> Working with variables in Cmake gets tricky because you don't know quite which order things will happen in. 
> I opted for KConfig string that can contain Cmake variables. 

Why is this less tricky ?
Now you are  depending on Kconfig processing order.

You are depending upon the fact that this line:
https://github.com/nrfconnect/sdk-nrf/blob/568fb2fa45e129927f5d148877e788ef285ce1fa/subsys/tfm/Kconfig#L17-L19

must be processed before this line:
https://github.com/nrfconnect/sdk-zephyr/blob/912208a6d5fef1e01bc1813efed76b615140e1ac/modules/Kconfig.tfm#L23
```
config TFM_BOARD_BASE_DIR
	string
	default "${TFM_BOARD_BASE_DIR_DEFAULT}" 
```

and do you know how the process order Zephyr module Kconfig files are ensured compared to Zephyr module CMakeLists.txt files ?
They are processed in exactly same order, so that if `moduleA` is processed before `moduleB` in Kconfig, then the same is true for the `moduleA` and `moduleB` in CMake. (Of course for Zephyr modules having both a CMakeLists.txt and Kconfig file)

The Kconfig process order of modules is placed in `<build>/Kconfig/Kconfig.modules`:
```
osource "/projects/github/ncs/nrf/Kconfig.nrf"
osource "/projects/github/ncs/bootloader/mcuboot/zephyr/Kconfig"
osource "/projects/github/ncs/nrfxlib/Kconfig.nrfxlib"
```
and thereafter all Kconfig files in `$ZEPHYR_BASE/modules/Kconfig.<module>`.

and CMake in `<build>/zephyr_modules.txt`:
```
"nrf":"/projects/github/ncs/nrf":"/projects/github/ncs/nrf"
"mcuboot":"/projects/github/ncs/bootloader/mcuboot":"/projects/github/ncs/bootloader/mcuboot/zephyr"
...
"nrfxlib":"/projects/github/ncs/nrfxlib":"/projects/github/ncs/nrfxlib"
"tfm":"/projects/github/ncs/modules/tee/tfm":"/projects/github/ncs/modules/tee/tfm"
...
```

and this order is same order as `west list` adjusted for any `depends` dependencies, so that those dependencies are guaranteed.

This means the manifest project is first in list, then local defined projects, then imported projects.
Unless a module is using `depends`, which means it will of course be processed after the module on which it depends.
https://docs.zephyrproject.org/latest/guides/modules.html#zephyr-module-dependencies

So if this is your only concern, I would stick to the CMake, and avoid the additional patches.

Note: the fact that all Kconfig files in `$ZEPHYR_BASE/modules/Kconfig.<module>` are processed in a bunh, is because they are not truly mdule code, which is one of the reasons for fixing this as part of: https://github.com/zephyrproject-rtos/zephyr/pull/30904

Comment (oyvindronningstad) [reply subsys/partition_manager/Kconfig]: I get your point about resolving order. But is modifying a CONFIG_* variable in Cmake really the best way to solve this? That strikes me as hacky :P.

Comment (tejlmand) [reply subsys/partition_manager/Kconfig]: > > That strikes me as hacky :P.

I agree on this, and as I mentioned, I believe  that we should aim at pushing a proper solution to TFM, as I wrote earlier:
> I think it would be even better if TFM itself could be improved to allow out of tree boards.

then the corresponding in Zephyr would not need to be a path, but could be a choice or similar config that would translate into a TFM board dir / root to be passed on to TFM build system.
Such a setting translation could then be defined by the module itself, in this case nrf.

With such a goal, I think a pure NCS local fix / hack of proposed nature would introduce the minimal amount of changes locally, and require no upstream patches. Patches that would anyway have to be re-worked when TFM truly supports out-of-tree boards.

Comment (oyvindronningstad) [reply subsys/partition_manager/Kconfig]: Ok, I'm working on a proposal for upstream where it's possible to pass a platform as an absolute path. Do you see any problems with that?

Comment (tejlmand) [reply subsys/partition_manager/Kconfig]: > Do you see any problems with that?

Guess I would need to see the proposal, in order to tell :wink: 
But it sounds like a good idea.

Comment (oyvindronningstad) [reply subsys/partition_manager/Kconfig]: Ok, cool. Take a look: https://review.trustedfirmware.org/c/TF-M/trusted-firmware-m/+/8064

File: CODEOWNERS (hunk 18 above)
[the remark this answers is not in the document]
Comment (oyvindronningstad) [reply CODEOWNERS:122]: Moved to modules/tfm

File: subsys/partition_manager/pm.yml.tfm (hunk 7 above)
Comment (hakonfam) [reply subsys/partition_manager/pm.yml.tfm:36]: No, it makes sense to me now :)

File: subsys/partition_manager/Kconfig (hunk 17 above)
Comment (oyvindronningstad) [reply subsys/partition_manager/Kconfig]: This is now merged and brought into https://github.com/nrfconnect/sdk-trusted-firmware-m/pull/1

File: subsys/tfm/boards/nrf5340_cpuapp/CMakeLists.txt (hunk 13 above)
Comment (oyvindronningstad) [reply subsys/tfm/boards/nrf5340_cpuapp/CMakeLists.txt]: Done

File: subsys/tfm/boards/nrf5340_cpuapp_eng_a/CMakeLists.txt (hunk 12 above)
Comment (oyvindronningstad) [reply subsys/tfm/boards/nrf5340_cpuapp_eng_a/CMakeLists.txt]: Done

File: subsys/tfm/boards/nrf9160/CMakeLists.txt (hunk 14 above)
[the remark this answers is not in the document]
Comment (oyvindronningstad) [reply subsys/tfm/boards/nrf9160/CMakeLists.txt]: Thanks, done

**Comment by oyvindronningstad** [comment]
@tejlmand all your comments should be addressed now.

File: modules/tfm/Kconfig [hunk 22]
```diff
@@ -0,0 +1,15 @@
...
+# Copyright (c) 2021 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+#
+
+config TFM_BOARD
+	string
+	default "${ZEPHYR_NRF_MODULE_DIR}/modules/tfm/boards/nrf9160" if SOC_NRF9160_SICA
+	default "${ZEPHYR_NRF_MODULE_DIR}/modules/tfm/boards/nrf5340_cpuapp_eng_a" if SOC_NRF5340_CPUAPP_QKAA && NRF5340_CPUAPP_ERRATUM19
+	default "${ZEPHYR_NRF_MODULE_DIR}/modules/tfm/boards/nrf5340_cpuapp" if SOC_NRF5340_CPUAPP_QKAA && !NRF5340_CPUAPP_ERRATUM19
```
Comment (tejlmand) [inline modules/tfm/Kconfig:11]: @ioannisg whatever happened to this:
> I did not read the whole conversation but I lean towards accepting the latest proposal. I'll open a PR and we can finalize the discussion there

https://github.com/zephyrproject-rtos/zephyr/pull/30062#discussion_r526716896

I think now is a good time to actually cleanup this, and remove the use of paths from Kconfig.tfm.

Depending on the order in which we take in PRs, I can also add this cleanup as part of this PR: https://github.com/zephyrproject-rtos/zephyr/pull/30905

Is this PR expected to enter 1.5 ? (it's not milestoned)

Comment (tejlmand) [reply modules/tfm/Kconfig:11]: guess this will happen as part of this: https://github.com/zephyrproject-rtos/zephyr/pull/30145
and thus probably waits for this: https://github.com/zephyrproject-rtos/zephyr/pull/30905

**Review by tejlmand — APPROVED** [review]

**Review by tejlmand — APPROVED** [review]

File: subsys/tfm/boards/board/RTE_Device.h (hunk 1 above)
Comment (oyvindronningstad) [reply subsys/tfm/boards/board/RTE_Device.h]: @anangl?

File: modules/tfm/boards/CMakeLists.txt [hunk 23]
```diff
@@ -0,0 +1,36 @@
+#-------------------------------------------------------------------------------
+# Copyright (c) 2021, Nordic Semiconductor ASA.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+#-------------------------------------------------------------------------------
+
+# NOTE: All code in this folder and subfolders is run as part of the TF-M
+#       build, not in a Zephyr context.
```
Comment (hakonfam) [inline modules/tfm/boards/CMakeLists.txt:9]: Should we note the exact file which includes this file? That could be helpful for debugging.

**Review by hakonfam — APPROVED** [review]

File: modules/tfm/boards/CMakeLists.txt (hunk 23 above)
Comment (oyvindronningstad) [reply modules/tfm/boards/CMakeLists.txt:9]: I think such a comment is likely to become stale. We could instead explain it in docs. What do you think?

**Review by ioannisg — APPROVED** [review]

File: modules/tfm/boards/CMakeLists.txt (hunk 23 above)
Comment (ioannisg) [reply modules/tfm/boards/CMakeLists.txt:9]: I think docs is a better place.
